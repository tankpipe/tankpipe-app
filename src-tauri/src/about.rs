use serde::{Serialize, Deserialize};
use data_encoding::BASE64;


const VERSION: &str = env!("CARGO_PKG_VERSION");
const CONTACT: &str = "aHR0cHM6Ly9kaXNjb3JkLmdnL3V6cTg5R1ByVWc=";

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub struct About {
    pub version: String,
    pub contact: String,
}

impl About {
    pub fn new() -> About {
        let input: Vec<u8> = CONTACT.into();
        let binding = BASE64.decode(&input).unwrap();
        let s = String::from_utf8_lossy(&binding);
        About{
            version: VERSION.to_string(),
            contact: s.into_owned()
        }
    }
}
