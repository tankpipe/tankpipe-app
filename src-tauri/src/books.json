{   
    "accounts": {
      "44742ea3-f835-4267-b62e-0c1d4796d575": {
        "id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "name": "Savings Account 1",
        "account_type": "Debit",
        "balance": "0",
        "starting_balance": "0"
      },
      "66ac7995-4991-495a-b44e-d9882d579260": {
        "id": "66ac7995-4991-495a-b44e-d9882d579260",
        "name": "Credit Account 1",
        "account_type": "Credit",
        "balance": "0",
        "starting_balance": "0"
      }
    },
    "scheduled_transactions": [
      {
        "id": "8565ba58-2df8-4690-b4f1-7849b2b4b48f",
        "name": "Some income",
        "period": "Months",
        "frequency": 1,
        "start_date": "2022-06-04",
        "last_date": "2022-06-04",
        "amount": "200",
        "description": "Money in",
        "dr_account_id": "cb946148-79b0-43bc-a3b8-b8bd811aa1ad",
        "cr_account_id": "c77327ef-d27d-434c-a205-2e975e822c87"
      }
    ],
    "transactions": [
      {
        "id": "4ed6200e-4d96-4c40-b140-36de0cfc7bc8",
        "date": "2022-06-04",
        "description": "received moneys",
        "dr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "cr_account_id": "66ac7995-4991-495a-b44e-d9882d579260",
        "amount": "10000",
        "status": "Recorded"
      },
      {
        "id": "f70e1b5d-12ee-47b0-9fbb-ab7671ff4595",
        "date": "2022-06-05",
        "description": "Gave some moneys back",
        "dr_account_id": "66ac7995-4991-495a-b44e-d9882d579260",
        "cr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "amount": "98.99",
        "status": "Recorded"
      },
      {
        "id": "4ed6200e-4d96-4c40-b140-36de0cfc7bc9",
        "date": "2022-07-01",
        "description": "received more moneys",
        "dr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "cr_account_id": "66ac7995-4991-495a-b44e-d9882d579260",
        "amount": "250",
        "status": "Recorded"
      },
      {
        "id": "4ed6200e-4d96-4c40-b140-36de0cfc7bd1",
        "date": "2022-07-01",
        "description": "received more moneys",
        "dr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "cr_account_id": null,
        "amount": "150.45",
        "status": "Recorded"
      }
    ]
  }