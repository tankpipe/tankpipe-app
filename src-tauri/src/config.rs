use std::{ffi::{OsString, OsStr}, path::PathBuf};
use serde::{Deserializer, Serializer, Serialize, Deserialize};

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub struct FileDetails {
    pub name: String,
    #[serde(serialize_with = "serialize_osstring")]
    #[serde(deserialize_with = "deserialize_osstring")]
    pub path: OsString,
}

impl FileDetails {
    pub fn empty() -> FileDetails {
        FileDetails{ name: "".to_string(), path: OsString::from("") }
    }

    pub fn from_path(name: &str, path: PathBuf) -> FileDetails {
        FileDetails{ name: name.to_string(), path: path.into_os_string() }
    }

    pub fn from_os_string(name: &str, path: OsString) -> FileDetails {
        FileDetails{ name: name.to_string(), path: path }
    }
}

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub enum DateFormat {
    Locale,
    Regular,
    US,
    ISO,
}

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub struct Config {
    pub version: String,
    #[serde(serialize_with = "serialize_osstring")]
    #[serde(deserialize_with = "deserialize_osstring")]
    pub data_dir: OsString,
    #[serde(serialize_with = "serialize_osstring")]
    #[serde(deserialize_with = "deserialize_osstring")]
    pub config_dir: OsString,
    pub last_file: FileDetails,
    pub recent_files: Vec<FileDetails>,
    pub display_date_format: DateFormat,
    pub import_date_format: String
}

impl Config {
    pub fn data_dir_as_os_str(&self) -> &OsStr {
        &self.data_dir.as_os_str()
    }

    pub fn settings_path(&self) -> PathBuf {
        PathBuf::from(self.config_dir.clone()).join("settings.json")
    }

    pub fn file_path(&self, name: &str) -> PathBuf {
        PathBuf::from(self.data_dir.clone()).join(name)
    }

    pub fn set_last_from_path(&mut self, path: OsString, name: &str) {
        let last_file: FileDetails;
        let os_string_path = OsString::from(path);
        self.last_file.path = os_string_path.clone();
        self.last_file.name = name.to_string();
        let index = self.recent_files.iter().position(|f| *f.path == os_string_path);
        match index {
            Some(i) => {
                last_file = self.recent_files.get(i).unwrap().clone();
                self.recent_files.remove(index.unwrap());
            },
            None =>last_file = FileDetails::from_os_string(name, os_string_path)
        }
        self.recent_files.insert(0, last_file);
    }

    pub fn set_last(&mut self, last_file: FileDetails) {
        if let Some(index) = self.recent_files.iter().position(|f| *f.path == last_file.path) {
            self.recent_files.remove(index);
        }
        self.last_file = last_file.clone();
        self.recent_files.insert(0, last_file);
    }

}

pub fn deserialize_osstring<'de, D>(deserializer: D) -> Result<OsString, D::Error>
    where D: Deserializer<'de>
{
    let path = String::deserialize(deserializer)?;
    Ok(OsString::from(path))
}

pub fn serialize_osstring<S>(path: &OsString, serializer: S) -> Result<S::Ok, S::Error>
    where S: Serializer,
{
    serializer.serialize_some(&PathBuf::from(path.as_os_str()))
}