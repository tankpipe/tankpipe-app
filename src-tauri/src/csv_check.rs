use serde::{Serialize, Deserialize};


use crate::reader::ColumnTypes;

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct CsvCheck {
    pub column_types: ColumnTypes,
    pub header_row: Vec<String>,
    pub sample_rows: Vec<Vec<String>>,
    pub mapping_exists: bool,
}

impl CsvCheck {
    pub fn create_new(column_types: ColumnTypes, header_row: Vec<String>, sample_rows: Vec<Vec<String>>) -> CsvCheck {
        CsvCheck {
            column_types: column_types,
            header_row: header_row,
            sample_rows: sample_rows,
            mapping_exists: false,
        }
    }

    pub fn set_mapping_exists(self, mapping_exists: bool) -> CsvCheck {
        CsvCheck {
            column_types: self.column_types,
            header_row: self.header_row,
            sample_rows: self.sample_rows,
            mapping_exists: mapping_exists,
        }
    }
}


