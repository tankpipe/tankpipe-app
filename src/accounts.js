import { writable } from 'svelte/store'

const accounts = writable({})

export {accounts as accounts}