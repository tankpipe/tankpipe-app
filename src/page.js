import { writable } from 'svelte/store'

const views = {
    BOOKS: "BOOKS",
    ACCOUNTS: "ACCOUNTS",
    TRANSACTIONS: "TRANSACTIONS",
    SCHEDULES: "SCHEDULES",
    SETTINGS: "SETTINGS",
    NET_ASSETS: "NET_ASSETS",
    JOURNAL: "JOURNAL",
}

const modes = {
    NEW: "NEW",
    EDIT: "EDIT",
    MULTI_EDIT: "MULTI_EDIT",
    LIST: "LIST",
    LOAD: "LOAD"
}

const page = writable({view: views.ACCOUNTS, mode:modes.LIST, payload:{}})

const isEditMode = (page) => {
    return page.mode === modes.EDIT || page.mode === modes.NEW || page.mode === modes.MULTI_EDIT
}

const isSingleEditMode = (page) => {
    return page.mode === modes.EDIT || page.mode === modes.NEW
}

const isMultiEditMode = (page) => {
    return page.mode === modes.MULTI_EDIT
}

const isListMode = (page) => {
    return page.mode === modes.LIST
}

export {page, views, modes, isEditMode, isMultiEditMode, isSingleEditMode, isListMode}