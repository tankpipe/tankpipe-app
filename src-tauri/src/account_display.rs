use accounts::account::{AccountType, Account};
use chrono::NaiveDate;
use rust_decimal::Decimal;
use serde::{Serialize, Deserialize};
use uuid::Uuid;
use accounts::serializer::*;
use rust_decimal_macros::dec;

#[derive(Clone, Serialize, Deserialize)]
pub struct NewAccount {
	pub name: String,
	pub account_type: AccountType,
	pub starting_balance: Decimal
}

impl NewAccount {
	pub fn to_account(self) -> Account {
		Account {
			id: Uuid::new_v4(),
			name: self.name,
			starting_balance: self.starting_balance,
			account_type: self.account_type,
            balance: dec!(0)
		}
	}

}


#[derive(Clone, Serialize, Deserialize)]
pub struct DateParam {
    #[serde(serialize_with = "serialize_naivedate")]
    #[serde(deserialize_with = "deserialize_naivedate")]
	pub date: NaiveDate
}

