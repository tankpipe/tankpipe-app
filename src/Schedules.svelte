<script>
	import EditSchedule from './EditSchedule.svelte'
	import Icon from '@iconify/svelte'
    import { page, isEditMode, views, modes } from './page';

	export let accounts = []
	let curSchedule

	const close = () => {
        console.log("close")
        page.set({view: views.SCHEDULES, mode: modes.LIST})
		loadSchedules()
    }

	$: {
		loadSchedules()
    }

	const selectSchedule = (transaction) => {
		curSchedule = transaction
		page.set({view: views.SCHEDULES, mode: modes.EDIT})
	}

	let schedules = [];
	const loadSchedules = async () => {
		console.log("loadSchedules");
   		schedules = await invoke('schedules');
	};

	const formatter = new Intl.NumberFormat('en-AU', {
		minimumFractionDigits: 2,
		maximumFractionDigits: 2,
	});

	const matchAccount = (accountId) =>  {
        if (!accountId) return null
        let match = accounts.filter(a => a.id == accountId)
        return match.length > 0 ? match[0] : null
    }

	const getAccountName = (accountId) => {
		const account = matchAccount(accountId)
		return account ? account.name : "None"
	}

	const handleAddClick = () => {
		page.set({view: views.SCHEDULES, mode: modes.NEW})
	}

</script>

<div class="account-heading">
	{#if !isEditMode($page)}
	<div class="toolbar"><div class="toolbar-icon" on:click={handleAddClick} title="Create a new schedule"><Icon icon="mdi:plus-box-outline"  width="24"/></div></div>
	<div class="form-heading">Schedules</div>
	{/if}
</div>
{#if isEditMode($page)}
<EditSchedule {close} {accounts} {curSchedule} {loadSchedules}/>
{/if}
{#if !isEditMode($page)}
<div class="scroller">
	{#if schedules.length < 1}
	<div class="message">No schedules</div>
	{/if}
		{#each schedules as s}
			<!-- svelte-ignore a11y-click-events-have-key-events -->
			<div class="card" on:click={() => selectSchedule(s)}>
				<div class="row">
					<div class="widget">
						<div class="description">{s.name}</div>
					</div>
				</div>
				<hr/>
				<div class="row">
					<div class="widget schedule-entries">
					<table>
						{#each s.entries as e}
						<tr>
							<td>{getAccountName(e.account_id)}</td>
							<td>{#if e.transaction_type == "Debit"} <div class="money">{formatter.format(e.amount)}</div>{/if}</td>
							<td>{#if e.transaction_type == "Credit"} <div class="money">{formatter.format(e.amount)}</div>{/if}</td>
						</tr>
						{/each}
					</table>
				</div>
				</div>
				<div class="row">
					<div class="widget">
						<div class="label">Last</div>
						<div class="account">{s.last_date == "null"?"-":s.last_date}</div>
					</div>
				</div>
			</div>
		{/each}
</div>
{/if}

<style>
	.widget {
		display: inline-block;
		text-align: left;
		margin: 10px 10px;
		color: #F0F0F0;
		vertical-align: top;
	}

	.schedule-entries {
		font-size: .9em;
		color: #C0C0C0;
	}

	.row {
		display: block;
		text-align: left;
	}

	.label {
		font-size: .8em;
		color: #aaa !important;
		margin-bottom: 5px;
	}

	.scroller {
		clear: both;
	}

	.card {
		float: left;
		clear: both;
		margin: 10px;
		background-color: #524e4e;
		padding: 5px;
		border-radius: 10px;
	}

	.card:hover {
		cursor: pointer;
	}

	.money {
		text-align: right !important;
		min-width: 100px;
	}

	.description {
		min-width: 500px;
		white-space: nowrap;
		font-weight: bold;
	}

	.account {
		min-width: 200px;
		white-space: nowrap;
	}

	hr {
		border: 1px solid #444;
		margin: 0 -5px;
	}

	.message {
		color: #EFEFEF;
		margin-bottom: 20px;
		text-align: left;
		background-color: #303030;
		padding:10px;
		border-radius: 10px;
	}

	.toolbar {
		float: right;
		color: #C0C0C0;
		margin-bottom: 10px;
	}

	.toolbar-icon:hover{
		color: #F0F0F0;
		cursor: pointer;
	}

</style>