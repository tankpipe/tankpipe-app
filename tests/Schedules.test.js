import { render } from '@testing-library/svelte'
import Schedules from '../src/Schedules.svelte'
import {page, views, modes} from '../src/page'
import schedules_data from './data/schedules_data.json'
import {accounts} from '../src/accounts.js'
import account_data from './data//account_data.json'
import { mockIPC } from "@tauri-apps/api/mocks"

accounts.set(account_data)

it('is displayed correctly', async () => {
    page.set({view: views.SCHEDULES, mode: modes.LIST})
    mockIPC((cmd, args) => {
        switch (cmd) {
            case "schedules": return schedules_data
        }

    });

    const {findByText, container} = render(Schedules)
    const _waitForRenderUpdate = await findByText('Test Schedule 1')
    expect(container.outerHTML).toMatchSnapshot();
});

