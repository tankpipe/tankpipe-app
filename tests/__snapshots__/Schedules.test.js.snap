// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`is displayed correctly 1`] = `"<div><div class="account-heading"><div class="toolbar svelte-wfwt0p"><div class="toolbar-icon svelte-wfwt0p" title="Create a new schedule"></div></div> <div class="form-heading">Schedules</div></div>  <div class="controls svelte-wfwt0p"><div class="form-row2 svelte-wfwt0p"><div class="widget svelte-wfwt0p"><div class="label label-column svelte-wfwt0p">Schedule until </div><div class="date-input field svelte-wfwt0p"><input type="date" class="svelte-wfwt0p"></div></div></div></div> <div class="scroller svelte-wfwt0p"> <div class="card svelte-wfwt0p"><div class="row svelte-wfwt0p"><div class="widget svelte-wfwt0p"><div class="description svelte-wfwt0p">Test Schedule 1</div></div></div> <hr class="svelte-wfwt0p"> <div class="row svelte-wfwt0p"><div class="widget schedule-entries svelte-wfwt0p"><table><tr><td>Account 2</td> <td><div class="money svelte-wfwt0p">200.00</div></td> <td></td> </tr><tr><td>Account 1</td> <td></td> <td><div class="money svelte-wfwt0p">200.00</div></td> </tr></table></div></div> <div class="row svelte-wfwt0p"><div class="widget svelte-wfwt0p"><div class="label svelte-wfwt0p">Last</div> <div class="account svelte-wfwt0p">-</div></div></div> </div><div class="card svelte-wfwt0p"><div class="row svelte-wfwt0p"><div class="widget svelte-wfwt0p"><div class="description svelte-wfwt0p">Test Schedule 2</div></div></div> <hr class="svelte-wfwt0p"> <div class="row svelte-wfwt0p"><div class="widget schedule-entries svelte-wfwt0p"><table><tr><td>Account 1</td> <td><div class="money svelte-wfwt0p">50.00</div></td> <td></td> </tr><tr><td>Account 2</td> <td></td> <td><div class="money svelte-wfwt0p">50.00</div></td> </tr></table></div></div> <div class="row svelte-wfwt0p"><div class="widget svelte-wfwt0p"><div class="label svelte-wfwt0p">Last</div> <div class="account svelte-wfwt0p">-</div></div></div> </div></div></div>"`;
